clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -2622,36 +2649,113 @@ raw2trace_t::thread_file_at_eof(raw2trace_thread_data_t *tdata)
+            app_pc branch_addr = reinterpret_cast<app_pc>(entry.addr);
+            trace_type_t branch_type = static_cast<trace_type_t>(entry.type);
+            if (next_instr_pc == nullptr &&
+                (target == nullptr || entry.type == TRACE_TYPE_INSTR_CONDITIONAL_JUMP)) {
+                // This is a trace-final or window-final branch but we do not have
+                // its taken/target without a subsequent instr: just delete it.
+                DEBUG_ASSERT(instr_index == instr_count - 1);
+                int erase_from = i;
+                while (erase_from > 0 &&
+                       tdata->delayed_branch[erase_from - 1].type ==
+                           TRACE_TYPE_ENCODING) {
+                    log(4, "Erasing cached encoding for %p\n",
+                        tdata->delayed_branch_decode_pcs[instr_index]);
+                    tdata->encoding_emitted.erase(
+                        tdata->delayed_branch_decode_pcs[instr_index]);
+                    --erase_from;
+                }
+                VPRINT(
+                    4,
+                    "Discarded %zd entries for final branch without subsequent instr\n",
+                    tdata->delayed_branch.size() - erase_from);
+                tdata->delayed_branch.erase(tdata->delayed_branch.begin() + erase_from,
+                                            tdata->delayed_branch.end());
+                tdata->delayed_branch_decode_pcs.pop_back();
+                tdata->delayed_branch_target_pcs.pop_back();
+                break;
             } else {
+                if (target == nullptr) {
+                    DEBUG_ASSERT(!type_is_instr_direct_branch(
+                        static_cast<trace_type_t>(entry.type)));
+                    trace_entry_t local[3];
+                    int size = trace_metadata_writer_t::write_marker(
+                        reinterpret_cast<byte *>(local), TRACE_MARKER_TYPE_BRANCH_TARGET,
+                        reinterpret_cast<uintptr_t>(next_instr_pc));
+                    DEBUG_ASSERT(static_cast<size_t>(size) <= sizeof(local));
+                    for (int local_idx = 0;
+                         local_idx < size / static_cast<int>(sizeof(local[0]));
+                         ++local_idx) {
+                        tdata->delayed_branch.insert(tdata->delayed_branch.begin() + i,
+                                                     local[local_idx]);
+                    }
+                    VPRINT(4, "Inserted indirect branch target %p\n", next_instr_pc);
+                } else if (entry.type == TRACE_TYPE_INSTR_CONDITIONAL_JUMP) {
+                    if (target == next_instr_pc) {
+                        branch_type = TRACE_TYPE_INSTR_TAKEN_JUMP;
+                    } else {
+                        branch_type = TRACE_TYPE_INSTR_UNTAKEN_JUMP;
+                    }
+                    entry.type = static_cast<unsigned short>(branch_type);
+                }
```

Comment:
else ASSERT(false)?

Reply:
But there are other types of branches which we pass through here: unconditional direct branches: TRACE_TYPE_INSTR_DIRECT_CALL and TRACE_TYPE_INSTR_DIRECT_JUMP.